File: analysis/queen/queen.py
import termtables as tt
global count
count=0
def canPlace(k:int,i:int,result:list[int]):
    for j in range(0,k):
        if(result[j]==i or abs(result[j]-i)==abs(j-k)):
            return False
    return True

def printResult(result:list[int]):
    temp = result.copy()
    for i in range(0,len(temp)):
        temp[i]+=1
    print(temp)

def draw(result:list[int]):
    board = [['']*(len(result)) for i in range(len(result))]
    for i in range(len(result)):
        board[i][result[i]]='♕'
    tt.print(board,style=tt.styles.thick)

def queens(k:int,n:int,result:list[int]):
    global count
    for i in range(0,n):
        if(canPlace(k,i,result)):
            result[k]=i
            if(k==n-1): 
                printResult(result)
                # draw(result)
                count+=1
            else: queens(k+1,n,result)

if(__name__=="__main__"):
    n=int(input('Enter Number of Queens: '))
    result=[0]*n
    queens(0,n,result)
    print(f'Number of solutions: {count}')

File: analysis/queen/queen.cpp
#include<iostream>
#include<vector>
using namespace std;
int count=0;

bool canPlace(int k,int i,vector<int>& result){
    for(int j=0;j<result.size();++j){
        if(result[j]==i or abs(result[j]-i)==abs(j-k)){
            return false;
        }
    }
    return true;
}

void printResult(vector<int> result){
    for(int i=0;i<result.size();i++){
        result.at(i)++;
        cout<<result.at(i)<<' ';
    }
    cout<<'\n';
    // cout>>result>>endl;
}

void queens(int k,int n,vector<int> result){
    for(int i=0;i<result.size();++i){
        if(canPlace(k,i,result)){
            result[k]=i;
            if(k==n-1){
                printResult(result);
                count++;
            }
            else{
                queens(k+1,n,result);
            }
        }
    }
}

int main(){
    int n;
    cout<<"Enter number of queens: ";
    cin>>n;
    int k=0;
    vector<int> result;
    result.reserve(n);
    for(int i=0;i<result.size();i++){
        result.at(i)=0;
    }
    queens(k,n,result);
    cout<<"\nNumber of solutions are: "<<count<<endl;
}